fjord/heartbeat/healthchecks.py

```diff
@@ -0,0 +1,205 @@
+
+
+@register_check
+class CheckAnyAnswers(Check):
+    """Are there any heartbeat answers? If not, that's very bad."""
+    name = 'any answers check'
+
+    @classmethod
+    def check(cls):
+        day_ago = datetime.now() - timedelta(days=1)
+        count = Answer.objects.filter(received_ts__gt=day_ago).count()
+
+        if count == 0:
+            return Result(
+                cls.name,
+                SEVERITY_HIGH,
+                '0 answers in last 24 hours.',
+                str(count)
+            )
+        return Result(
+            cls.name,
+            SEVERITY_LOW,
+            '%s answers in last 24 hours.' % str(count),
+            str(count)
+        )
+
+
+def tableify(table):
+    num_cols = 0
+    maxes = []
+
+    for row in table:
+        num_cols = max(num_cols, len(row))
+        if len(maxes) < len(row):
+            maxes.extend([0] * (len(row) - len(maxes)))
+        for i, cell in enumerate(row):
+            maxes[i] = max(maxes[i], len(str(cell)))
+
+    def fix_row(maxes, row):
+        return '  '.join([
+            str(cell) + (' ' * (maxes[i] - len(str(cell))))
+            for i, cell in enumerate(row)
+        ])
+
+    return '\n'.join(
+        [
+            fix_row(maxes, row)
+            for row in table
+        ]
+    )
+
+
+@register_check
+class CheckMissingVotes(Check):
+    """should freak out if votes are 0 for any 'large' cell."""
+    name = 'missing votes check'
+
+    @classmethod
+    def check(cls):
+        # Note: This SQL statement comes from Gregg. It's probably
+        # mysql-specific.
+        sql = """
+        SELECT
+            sum(score is not NULL) as nvoted,
+            DATE_FORMAT(received_ts, '%Y-%m-%d') as ydm,
+            version,
+            channel,
+            100*sum(flow_began_ts > 0) / count(received_ts) as pct_began,
+            100*sum(flow_offered_ts >0) / count(received_ts) as pct_offered,
+            100*sum(flow_voted_ts > 0)/ count(received_ts) as pct_voted,
+            100*sum(flow_engaged_ts > 0) / count(received_ts) as pct_engaged,
+            count(received_ts) as N
+        FROM heartbeat_answer
+        WHERE
+            received_ts > DATE_SUB(now(), interval 1 day)
+            AND is_test=0
+            AND survey_id="heartbeat-by-user-first-impression"
+            AND (locale='en-us')
+        GROUP BY version, channel, ydm
+        HAVING
+            N >= 50
+            and nvoted = 0
+        ORDER BY
+            channel,
+            version,
+            ydm;
+        """
+        cursor = connection.cursor()
+        cursor.execute(sql)
+        data = cursor.fetchall()
+
+        if not data:
+            # If we get nothing back, then we have serious issues.
+            return Result(
+                cls.name,
+                SEVERITY_HIGH,
+                'No data from query',
+                repr(data)
+            )
+
+        data = list(data)
+
+        # FIXME: What consistutes SEVERITY_HIGH here?
```

Comment:
I have no idea what "bad" looks like. I'll ask Gregg and he can let us know how to figure that algorithmically. Until then, I think it's ok to just say everything is "low" and print the output.